test/test_broker.hpp

```diff
@@ -1107,18 +1143,26 @@ class test_broker {
             mi::ordered_unique<
                 mi::tag<tag_client_id>,
                 BOOST_MULTI_INDEX_MEMBER(session_state, MQTT_NS::buffer, client_id)
+            >,
+            mi::ordered_unique<
```

Comment:
I am confused why you have converted the session_state to a shared pointer? Was the client id and connection shared-ptr not enough to uniquely identify a session?

Reply:
See https://github.com/redboltz/mqtt_cpp/pull/657/files#r476833703
You clear `con`. So I cannot use connection shared_ptr for non_active_sessions.

Reply:
You should use clientid to identify non-active session, since only one session (active or non-active) can exist for any specific clientid.

Reply:
Hmm... I considered using client_id.

1. A client CONNECT (ClientId=cid1, SessionExpiryInterval=100) to the broker
2. The broker added the client to active_sessions.
3. The client disconnected.
4. The broker removed the client from active_sessions and added to non_active_sessions.
5. The broker set timeout (100sec). "cid1" is captured.
6. The client CONNECT (ClientId=cid1, CleanStart=false, SessionExpiryInterval=0xffffffff) again.
7. The broker removed the client from non_active_sessions and added the new one to active_sessions.
8. Timer fired. Boost.Asio removed timer from timer management area and post io_context event. *1
9. session_state object is destroyed here and timer is cancelled in the destructor. But no matched timer.
10. The client disconnected.
11. The broker removed the client from active_sessions and added to non_active_sessions.
12. *1 event is processed and session_state is removed by the "cid1" even though the session has infinity lifetime.

Step 8 is controlled by Boost.Asio library. The timing is unpredictable. If Boost.Asio doesn't have internal thread, there is no interruption but I'm afraid that this assumption.

Step 8, 9 See https://stackoverflow.com/questions/43045192/how-to-avoid-firing-already-destroyed-boostasiodeadline-timer

I want to avoid unexpected match to new created session_state.
Using address is simpler way.

Reply:
I'm not sure the above scenario is actually happen. In the single thread environment it might never happen.
My point is dividing the issue.
Timer and weak_ptr approach works independently. Simply, unexpected match and/or accessing destructed object is never happen. 
I'd like to separate timer issue from MQTT client_id unique constraint.

Reply:
I think you are correct that in multithreaded environment, rapid connect, disconnect, connect, disconnect (similar to your example), can happen.

This would cause timer to potentially remove non-active-session that was for previous session.

Possible solutions:

1) Have timer check address of timer object in non-active session_state object to be removed, and compare to address of timer.

Problem, potentially same address re-used by allocator.

2) Some boost multi-index node ID of some sort?

Problem, may not exist, or if it exists, may be reused, so not unique.

3) Boost-multi-index intrusive pointer support? Either internal to boost-multi-index, or part of "session_state" struct.

4) Store boost-UUID in session-state object, copy uuid into timer, compare uuids before erasing.

5) This solution -- increased allocations, increased redirections caused by pointer access, instead of direct struct access.

6) Store mqtt_cpp::shared_scope_guard in session_state object, then store weak-ptr to shared_scope_guard in timer-lambda.

Benefits compared to 5)
Same number of allocations, but allocation of mqtt_cpp::shared_scope_guard is smaller than allocation of session_state.
Fewer pointer redirections. Quicker struct struct access.
Can improve test_broker code by moving session cleanup logic to the mqtt_cpp::shared_scope_guard.

7) This solution + change use of boost-multi-index so that other multi-indexes take advantage of the shared_ptr<session_state>. For example, maybe can improved saved subscriptions storage?

Reply:
Thank you for the comment. I will try 6.

Reply:
I updated the code. I found good approach. It is similar to 6. session_state is no longer shared_ptr. Only steady_clock tim_session_expiry is shared_ptr. It is default constructed at first and only when timer is required, then it is initialized by shared_ptr (make_shared is used). Timer is automatically cancelled via session_state destructor (shared_ptr dispose).
session_state pointer redirection is the same as the original (master code).